README


This archive contains the distribution CloudFlare-Client,
version v0.55.5:

  Object Orientated Interface to CloudFlare client API

This software is Copyright (c) 2016 by Peter Roberts.

This is free software, licensed under:

  The MIT (X11) License


This README file was generated by Dist::Zilla::Plugin::Readme v5.043.



Changes
Revision history for CloudFlare-Client

{{$NEXT}}
 - Use Const::Fast instead of Readonly
 - Throw string exceptions instead of objects (objects are overkill here)
 - Remove JSON::Any from tests

v0.55.5   2016-03-27
 - Travis build on previous release failed on build branch due to
   issues in [TravisYML]. Making another release to test fixes with
   Travis::TestRelease disabled

v0.55.4   2016-03-26
 - tweaks
 - fix commit message
 - test all the perls in Travis
 - change weaver.ini to work with changes to Test::Synopsis

v0.55.3   2015-05-01
 - really fix auto tagging...

v0.55.2   2015-04-30
 - fix auto git tagging on release

v0.55.1   2015-04-30
 - switch to a V-string

0.05_5    2015-04-27
 - use docker containers in TravisCI
 - update test targets in TravisCI
 - start providing a cpanfile
 - BREAKING: switch to AUTOLOAD dispatch with all arguments as a hash
 - tidy code into my latest coding style
 - no longer clean up pre-reqs list

0.05_4    2015-02-07
 - Change release tag format and travis stuff
 - Test::RequiresInternet for upstream existence check
 - Handle test dependencies better

0.05_3    2015-02-06
 - Optimise "feature" usages
 - Add build branch
 - JSON::Any is deprecated, use JSON::MaybeXS instead

0.05_2    2015-02-05
 - Drop auto date based prereq versions
 - Add TravisCI testing
 - Improve tests
 - Make all POD weaved
 - Rework dist.ini
 - Remove vestigial bovine

0.05_1    2015-02-03
 - Bump to more modern Perl (5.16)
 - Don't load unneeded dzil modules
 - Use ::Git::GatherDir
 - Move to GitHub
 - Fix cpan testers failure
 - bits and bobs
 - Update repo info - git repo split

0.04      2014-08-26
 - Bump to stable

0.03_9    2014-08-18
 - Reinstate more liberal dep pruning
 - Trim more cruft from pod
 - Add homepage to meta
 - Add no indirect 'fatal'
 - Add git meta correctly

0.03_8    2014-08-17
 - Make upstream test an author test and remove guards
 - Tidy imports
 - Immutiblise

0.03_7    2014-08-16
 - Moo -> Moose because CPAN tests are still failing

0.03_6    2014-08-15
 - Moo::Lax -> Moo because CPAN tests are still failing

0.03_5    2014-08-15
 - Unloosen dependency removal to see if it stops CPAN tests exploding

0.03_4    2014-08-14
 - No longer depend on Test::Moose or Moose
 - Use Moo::Lax in tests

0.03_3    2014-08-13
 - Loosen dependancy removal during builds to try and group more
 - Require newer Type::Utils because of a bug
 - Moo::Lax because we don't want fatal warnings

0.03_2    2014-08-12
 - Try to improve upstream test for CPAN testers
 - Tidy code a bit
 - Port to Moo
 - Port to use Type::Tiny

0.03_1    2014-08-08
 - Fix some exception tests on older Mooses by actually escaping brackets in
   regexes, should make Strawberry Perl happier
 - Try harder to test for HTTPS connectivity when testing upstream existence
   and make sure it is required for install
 - Method::Signatures -> Kavorka
 - Tidy distribution a bit

0.03_0    2014-08-02
 - Improve compatibility, in particular testing, for CPAN Testers
 - Tidy code a bit

0.03      2014-07-31
 - Add a test for upstream connectivity that won't break on CPAN testers
 - Perl requirement dropped to 5.014
 - Tidy code and make it more robust
 - Extensive testing of the API using Test::LWP::UserAgent
 - Require at least one zone as an argument to zoneCheck
 - Use Test::Exception rather than own stuff
 - Get rid of upstream testing, they fail at CPAN Testers

0.02      2014-07-16
 - Fix issues that prevent CPAN indexing
 - Package with Dist::Zilla

0.01      2014-07-10
 - First version, released on an unsuspecting world.


lib/CloudFlare/Client.pm
package CloudFlare::Client;

# ABSTRACT: Object Orientated Interface to CloudFlare client API

use strict;
use warnings;
use namespace::autoclean;
use Carp;
use Const::Fast;
use Moose;
use MooseX::StrictConstructor;
use Types::Standard 'Str';
use Try::Tiny;
no indirect 'fatal';

use LWP::UserAgent 6.02;
use LWP::Protocol::https 6.02;    # Not used but we want the dependency
use JSON::MaybeXS;

# VERSION

# CF credentials
has 'email' => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

has 'key' => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

const my $UA_STRING => "CloudFlare::Client/$VERSION";

sub _build_ua {
    my $ua = LWP::UserAgent->new;
    $ua->agent($UA_STRING);

    return $ua;
}
has '_ua' => (
    is       => 'ro',
    isa      => 'LWP::UserAgent',
    init_arg => undef,              # Can't be passed by user
    builder  => '_build_ua',
);

const my $BASE_URI => 'https://api.cloudflare.com/client/v4/';
const my %ALLOWED_REQUESTS => ( get => 1, post => 1, patch => 1, delete => 1 );

# Make a request to the API
sub request {
    my ( $self, $type, $end_point, $params ) = @_;

    my $lwp_method = lc $type;

    croak "Invalid request type $lwp_method, allowed types are " . join ' or ',
      keys %ALLOWED_REQUESTS
      unless exists $ALLOWED_REQUESTS{$lwp_method};

    # These are optional
    my @content_params =
      defined $params ? ( Content => encode_json($params) ) : ();

    my $res = $self->_ua->$lwp_method(
        $BASE_URI . $end_point,
        'X-Auth-Email' => $self->email,
        'X-Auth-Key'   => $self->key,
        'Content-Type' => 'application/json',    # Send this always, to be safe
        @content_params,
    );

    unless ( $res->is_success ) {
        my $content_type =
          defined $res->header('Content-Type')
          ? $res->header('Content-Type')
          : '';

        my $errors_info =
          $content_type eq 'application/json'
          ? ' and API reported error(s) '
          . encode_json( decode_json( $res->decoded_content )->{errors} )
          : '';

        croak 'HTTP request failed with status ', $res->status_line,
          $errors_info;
    }

    croak 'Only JSON responses are supported'
      unless $res->header('Content-Type') eq 'application/json';

    my $res_content = decode_json( $res->decoded_content );

    croak 'API reported error(s) ', encode_json( $res_content->{errors} )
      unless $res_content->{success};

    return $res_content;
}

__PACKAGE__->meta->make_immutable;
1;    # End of CloudFlare::Client

__END__

=for test_synopsis my ( $CF_USER, $CF_KEY, $ZONE, $INTERVAL );

=head1 SYNOPSIS

    use CloudFlare::Client;

    my $api = CloudFlare::Client->new(
        email => $CF_USER,
        key   => $CF_KEY
    );

    $api->stats( z => $ZONE, interval => $INTERVAL );

=head1 OVERVIEW

Please see the documentation at
L<https://www.cloudflare.com/docs/client-api.html> for information on the
CloudFlare client API and its arguments. API actions are mapped to methods of
the same name and arguments are passed in as a hash with keys as given in the
docs

Successful API calls return the response section from the upstream JSON
API. Failures for whatever reason throw exceptions under the
CloudFlare::Client::Exception:: namespace

=method new

Construct a new API object

    my $api = CloudFlare::Client->new(
        email => $CF_EMAIL,
        key   => $CF_API_KEY,
    );

=method request

Make a request to the API

    # $api->request( $HTTP_REQUEST_TYPE, $PATH_TO_RESOURCE [, $JSON_BODY] )
    $api->request( 'GET', 'zones' );

=head1 SEE ALSO

Mojo::Cloudflare
WebService::CloudFlare::Host

=head1 ACKNOWLEDGEMENTS

Thanks to CloudFlare providing an awesome free service with an API

=cut


lib/CloudFlare/Client/Exception/Upstream.pm
package CloudFlare::Client::Exception::Upstream;

# ABSTRACT: Upstream CloudFlare API Exception

use strict;
use warnings;
no indirect 'FATAL';
use namespace::autoclean;

use Readonly;
use Moose;
use MooseX::StrictConstructor;
use CloudFlare::Client::Types 'ErrorCode';

# VERSION

extends 'Throwable::Error';

has errorCode => (
    is  => 'ro',
    isa => ErrorCode,
);

__PACKAGE__->meta->make_immutable;
1;    # End of CloudFlare::Client::Exception::Upstream

__END__

=head1 SYNOPSIS

    use CloudFlare::Client::Exception::Upstream;

    CloudFlare::Client::Exception::Upstream->throw(
        message   => 'Bad things occured',
        errorCode => 'E_MAXAPI',
    );

    my $e = CloudFlare::Client::Exception::Upstream->new(
        message   => 'Bad things happened',
        errorcode => 'E_MAXAPI',
    );

    $e->throw;

=attr message

The error message thrown upstream, readonly

=attr errorCode

The error code thrown upstream, readonly. Valid values are undef,
E_UNAUTH, E_INVLDINPUT or E_MAXAPI. Readonly

=method throw

On the class, throw a new exception

    CloudFlare::Client::Exception::Upstream->throw(
        message   => 'Bad things occured',
        errorCode => 'E_MAXAPI',
    );

On an instance, throw that exception

    $e->throw;

=method new

Construct a new exception

    my $e = CloudFlare::Client::Exception::Upstream->new(
        message   => 'Bad things happened',
        errorcode => 'E_MAXAPI',
    );

=cut


lib/CloudFlare/Client/Exception/Connection.pm
package CloudFlare::Client::Exception::Connection;

# ABSTRACT: CloudFlare API Connection Exception

use strict;
use warnings;
no indirect 'fatal';
use namespace::autoclean;

use Readonly;
use Moose;
use MooseX::StrictConstructor;
use Types::Standard 'Str';

# VERSION

extends 'Throwable::Error';

has status => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

__PACKAGE__->meta->make_immutable;
1;    # End of CloudFlare::Client::Exception::Connection

__END__

=head1 SYNOPSIS

    use CloudFlare::Client::Exception::Connection;

    CloudFlare::Client::Exception::Connection->throw(
        message   => 'HTTPS connection failure',
        status    => '404',
    );

    my $e = CloudFlare::Client::Exception::Connection->new(
        message   => 'HTTPS connection failure',
        status    => '404',
    );
    $e->throw;

=attr message

The error message thrown upstream, readonly

=attr status

The status code for the connection failure, readonly

=method throw

On the class, throw a new exception

    CloudFlare::Client::Exception::Connection->throw(
        message   => 'HTTPS connection failure',
        status    => '404',
    );

On an instance, throw that exception

    $e->throw;

=method new

Construct a new exception

    my $e = CloudFlare::Client::Exception::Connection->throw(
        message   => 'HTTPS connection failure',
        errorcode => '404',
    );

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Object Orientated Interface to CloudFlare client API",
  "AUTHOR" => "Peter Roberts <me+dev\@peter-r.co.uk>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CloudFlare-Client",
  "LICENSE" => "mit",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "CloudFlare::Client",
  "PREREQ_PM" => {
    "JSON::MaybeXS" => 0,
    "Kavorka" => 0,
    "LWP::Protocol::https" => "6.02",
    "LWP::UserAgent" => "6.02",
    "Moose" => 0,
    "MooseX::StrictConstructor" => 0,
    "Readonly" => 0,
    "Throwable::Error" => 0,
    "Type::Library" => 0,
    "Type::Utils" => "0.039_12",
    "Types::Standard" => 0,
    "indirect" => 0,
    "mro" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "HTTP::Response" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "IPC::System::Simple" => 0,
    "JSON::Any" => 0,
    "Test::CPAN::Meta::JSON" => 0,
    "Test::Exception" => 0,
    "Test::LWP::UserAgent" => 0,
    "Test::Moose" => 0,
    "Test::More" => 0,
    "Test::RequiresInternet" => 0,
    "Test::TypeTiny" => 0,
    "Try::Tiny" => 0,
    "blib" => "1.01"
  },
  "VERSION" => "v0.55.5",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Spec" => 0,
  "HTTP::Response" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "IPC::System::Simple" => 0,
  "JSON::Any" => 0,
  "JSON::MaybeXS" => 0,
  "Kavorka" => 0,
  "LWP::Protocol::https" => "6.02",
  "LWP::UserAgent" => "6.02",
  "Moose" => 0,
  "MooseX::StrictConstructor" => 0,
  "Readonly" => 0,
  "Test::CPAN::Meta::JSON" => 0,
  "Test::Exception" => 0,
  "Test::LWP::UserAgent" => 0,
  "Test::Moose" => 0,
  "Test::More" => 0,
  "Test::RequiresInternet" => 0,
  "Test::TypeTiny" => 0,
  "Throwable::Error" => 0,
  "Try::Tiny" => 0,
  "Type::Library" => 0,
  "Type::Utils" => "0.039_12",
  "Types::Standard" => 0,
  "blib" => "1.01",
  "indirect" => 0,
  "mro" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
